socket/test_socket.c

```diff
@@ -0,0 +1,333 @@
+ *
+ * %LICENSE%
+ */
+
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
+#include "unity_fixture.h"
+
+
+static unsigned char data[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+
+TEST_GROUP(test_unix_socket);
+
+
+TEST_SETUP(test_unix_socket)
+{
+}
+
+
+TEST_TEAR_DOWN(test_unix_socket)
+{
+}
+
+
+TEST(test_unix_socket, unix_zero_len_send)
+{
+	int fd[3];
+	struct msghdr msg;
+	struct iovec iov;
+	struct cmsghdr *cmsg;
+	union {
+		char buf[CMSG_SPACE(sizeof(fd)) * 3];
+		struct cmsghdr align;
+	} u;
+	ssize_t n;
+
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) < 0)
+		FAIL("socketpair");
+
+	n = send(fd[0], data, 0, 0);
+	TEST_ASSERT_EQUAL_INT(n, 0);
+
+	n = sendto(fd[0], data, 0, 0, NULL, 0);
+	TEST_ASSERT_EQUAL_INT(n, 0);
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iov = NULL;
+	msg.msg_iovlen = 0;
+	n = sendmsg(fd[0], &msg, 0);
+	TEST_ASSERT_EQUAL_INT(n, 0);
+
+	memset(&msg, 0, sizeof(msg));
+	iov.iov_base = NULL;
+	iov.iov_len = 0;
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+	n = sendmsg(fd[0], &msg, 0);
+	TEST_ASSERT_EQUAL_INT(n, 0);
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 0;
+	msg.msg_control = u.buf;
+	msg.msg_controllen = CMSG_LEN(sizeof(int) * 2);
+	cmsg = CMSG_FIRSTHDR(&msg);
+	cmsg->cmsg_level = SOL_SOCKET;
+	cmsg->cmsg_type = SCM_RIGHTS;
+	cmsg->cmsg_len = CMSG_LEN(sizeof(int) * 2);
+	memcpy(CMSG_DATA(cmsg), fd, sizeof(int) * 2);
+	n = sendmsg(fd[0], &msg, 0);
+	TEST_ASSERT_EQUAL_INT(n, 0);
+
+	fd[2] = 33333;
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 0;
+	msg.msg_control = u.buf;
+	msg.msg_controllen = CMSG_LEN(sizeof(int) * 3);
+	cmsg = CMSG_FIRSTHDR(&msg);
+	cmsg->cmsg_level = SOL_SOCKET;
+	cmsg->cmsg_type = SCM_RIGHTS;
+	cmsg->cmsg_len = CMSG_LEN(sizeof(int) * 3);
+	memcpy(CMSG_DATA(cmsg), fd, sizeof(int) * 3);
+	n = sendmsg(fd[0], &msg, 0);
+	TEST_ASSERT_EQUAL_INT(n, -1);
+	TEST_ASSERT_EQUAL_INT(errno, EBADF);
+
+	close(fd[0]);
+	close(fd[1]);
+}
+
+
+TEST(test_unix_socket, unix_zero_len_recv)
+{
+	int fd[2];
+	struct msghdr msg;
+	struct iovec iov;
+	ssize_t n;
+
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, fd) < 0)
+		FAIL("socketpair");
+
+	n = recv(fd[1], NULL, 0, MSG_DONTWAIT);
+	TEST_ASSERT_EQUAL_INT(n, -1);
+	TEST_ASSERT_EQUAL_INT(errno, EAGAIN);
+
+	n = recvfrom(fd[1], NULL, 0, MSG_DONTWAIT, NULL, 0);
+	TEST_ASSERT_EQUAL_INT(n, -1);
+	TEST_ASSERT_EQUAL_INT(errno, EAGAIN);
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iov = NULL;
+	msg.msg_iovlen = 0;
+	n = recvmsg(fd[1], &msg, MSG_DONTWAIT);
+	TEST_ASSERT_EQUAL_INT(n, -1);
+	TEST_ASSERT_EQUAL_INT(errno, EAGAIN);
+
+	memset(&msg, 0, sizeof(msg));
+	iov.iov_base = NULL;
+	iov.iov_len = 0;
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+	n = recvmsg(fd[1], &msg, MSG_DONTWAIT);
+	TEST_ASSERT_EQUAL_INT(n, -1);
+	TEST_ASSERT_EQUAL_INT(errno, EAGAIN);
```

Comment:
Maybe test also for:
 -  non blocking when `fcntl(F_SETFL, O_NONBLOCK)`
 -  plain old `read`? :)